### PR TITLE
Add oceanalecn to workflow generator

### LINKED ISSUE #912: Recenter ens. members around the deterministic background.
The code/application exists (under utils) but is not implemented in the workflow yet. We need to pass recentering increments to the ensemble members.

### PR DESCRIPTION
**Pull request by AndrewEichmann-NOAA**
# Description

Adds task oceanalecn to the workflow generator, plus some necessary env var work to the jjob. The workflow generator adds oceanalecn only if ~~`nens > 0`~~ `self.do_hybvar`.

Resolves https://github.com/NOAA-EMC/GDASApp/issues/912

# Type of change

- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES
- Does this change require a documentation update? NO

# How has this been tested?

Example:

- Cycled test with and without ocnanalecen task on Hera

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by AndrewEichmann-NOAA** [comment]
FYI @guillaumevernieres @CoryMartin-NOAA

File: workflow/applications/gfs_cycled.py [hunk 1]
```diff
@@ -30,6 +30,10 @@ def __init__(self, conf: Configuration):
             elif eupd_cdump in ['gfs', 'gdas']:
                 self.eupd_cdumps = [eupd_cdump]
 
+        self.do_ocnanalecen = False
+        if self.nens > 0:
```
Comment (CoryMartin-NOAA) [inline workflow/applications/gfs_cycled.py]: I think it makes more sense to tie this into `self.do_hybvar`, no?

Comment (CoryMartin-NOAA) [reply workflow/applications/gfs_cycled.py]: I wouldn't even have these 3 lines added.

File: workflow/applications/gfs_cycled.py [hunk 2]
```diff
@@ -43,8 +47,10 @@ def _get_app_configs(self):
             configs += ['anal', 'analdiag']
 
         if self.do_jediocnvar:
-            configs += ['prepoceanobs', 'ocnanalprep', 'ocnanalbmat',
-                        'ocnanalrun', 'ocnanalchkpt', 'ocnanalpost']
+            configs += ['prepoceanobs', 'ocnanalprep', 'ocnanalbmat', 'ocnanalrun']
+            if self.do_ocnanalecen:
```
Comment (CoryMartin-NOAA) [inline workflow/applications/gfs_cycled.py]: As mentioned above, consider replacing this with `self.do_hybvar`

File: workflow/applications/gfs_cycled.py [hunk 3]
```diff
@@ -137,9 +143,10 @@ def get_task_names(self):
             gdas_gfs_common_tasks_before_fcst += ['anal']
 
         if self.do_jediocnvar:
-            gdas_gfs_common_tasks_before_fcst += ['prepoceanobs', 'ocnanalprep',
-                                                  'ocnanalbmat', 'ocnanalrun',
-                                                  'ocnanalchkpt', 'ocnanalpost']
+            gdas_gfs_common_tasks_before_fcst += ['prepoceanobs', 'ocnanalprep', 'ocnanalbmat', 'ocnanalrun']
+            if self.do_ocnanalecen:
```
Comment (CoryMartin-NOAA) [inline workflow/applications/gfs_cycled.py]: Same comment

File: workflow/applications/gfs_cycled.py (hunk 1 above)
Comment (guillaumevernieres) [reply workflow/applications/gfs_cycled.py]: > I think it makes more sense to tie this into `self.do_hybvar`, no?

I didn't like the use the `do_hybvar` switch for that task, so I suggested instead to link it to `nens` ... But I don't think this is the correct switch either. I'm fine with either solutions.

Comment (CoryMartin-NOAA) [reply workflow/applications/gfs_cycled.py]: pretty sure that if nens=0, do_hybvar is set to no. Is there a scenario where we will be running with atmospheric envar but only 3dvar ocean?

Comment (AndrewEichmann-NOAA) [reply workflow/applications/gfs_cycled.py]: removed

File: workflow/applications/gfs_cycled.py (hunk 2 above)
Comment (AndrewEichmann-NOAA) [reply workflow/applications/gfs_cycled.py]: changed

File: workflow/applications/gfs_cycled.py (hunk 3 above)
Comment (AndrewEichmann-NOAA) [reply workflow/applications/gfs_cycled.py]: changed

File: workflow/rocoto/gfs_tasks.py [hunk 4]
```diff
@@ -677,11 +677,38 @@ def ocnanalrun(self):
...
+                     }
+
+        task = rocoto.create_task(task_dict)
+
+        return task
+
+    def ocnanalchkpt(self):
+
+        deps = []
+        if self.app_config.do_ocnanalecen:
```
Comment (CoryMartin-NOAA) [inline workflow/rocoto/gfs_tasks.py]: need this fixed

Comment (AndrewEichmann-NOAA) [reply workflow/rocoto/gfs_tasks.py]: done

**Review by CoryMartin-NOAA — APPROVED** [review]

File: workflow/applications/gfs_cycled.py (hunk 1 above)
Comment (guillaumevernieres) [reply workflow/applications/gfs_cycled.py]: > pretty sure that if nens=0, do_hybvar is set to no. Is there a scenario where we will be running with atmospheric envar but only 3dvar ocean?

Yes, I could see a scenario where somebody wants to run the atmos hyb-envar without running the marine envar. But we can adjust/refine the switches later on as needed.

**Review by guillaumevernieres — APPROVED** [review]
:+1:

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests.

**Comment by guillaumevernieres** [comment]
> Conditionally approved pending successful completion of CI tests.

It's not yet tested as part of the g-w ci @WalterKolczynski-NOAA .

**Comment by WalterKolczynski-NOAA** [comment]
> > Conditionally approved pending successful completion of CI tests.
> 
> It's not yet tested as part of the g-w ci @WalterKolczynski-NOAA .

I'd still like to test it to make sure it doesn't break any existing tests.

@AndrewEichmann-NOAA Please merge in the current develop and then we can test this PR and get it merged.

**Comment by guillaumevernieres** [comment]
> > > Conditionally approved pending successful completion of CI tests.
> > 
> > 
> > It's not yet tested as part of the g-w ci @WalterKolczynski-NOAA .
> 
> I'd still like to test it to make sure it doesn't break any existing tests.
> 
> @AndrewEichmann-NOAA Please merge in the current develop and then we can test this PR and get it merged.

Could this PR be merged @WalterKolczynski-NOAA ?

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN [hunk 5]
```diff
@@ -13,7 +22,7 @@ source "${HOMEgfs}/ush/jjob_header.sh" -e "ocnanalecen" -c "base ocnanal ocnanal
 ###############################################################
 # Run relevant script
 
-EXSCRIPT=${GDASOCNCENPY:-${HOMEgfs}/scripts/exgdas_global_marine_analysis_ecen.py}
+EXSCRIPT=${GDASOCNCENPY:-${HOMEgfs}/ush/exgdas_global_marine_analysis_ecen.py}
```
Comment (aerorahul) [inline jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN]: This is an exscript and should be in `scripts/` not `ush/`

File: sorc/link_workflow.sh [hunk 6]
```diff
@@ -228,6 +228,7 @@ if [[ -d "${HOMEgfs}/sorc/gdas.cd/build" ]]; then
   ${LINK_OR_COPY} "${HOMEgfs}/sorc/gdas.cd/ush/ioda/bufr2ioda/run_bufr2ioda.py"    .
   ${LINK_OR_COPY} "${HOMEgfs}/sorc/gdas.cd/build/bin/imsfv3_scf2ioda.py"           .
   ${LINK_OR_COPY} "${HOMEgfs}/sorc/gdas.cd/scripts/exglobal_prep_ocean_obs.py"           .
+  ${LINK_OR_COPY} "${HOMEgfs}/sorc/gdas.cd/scripts/exgdas_global_marine_analysis_ecen.py"           .
```
Comment (aerorahul) [inline sorc/link_workflow.sh]: Should be linked in `scripts/` not `ush/`.
Same for Line 230.

File: .gitignore [hunk 7]
```diff
@@ -162,6 +162,7 @@ scripts/exemcsfc_global_sfc_prep.sh
 ush/chgres_cube.sh
 ush/emcsfc_ice_blend.sh
 ush/emcsfc_snow.sh
+ush/exgdas_global_marine_analysis_ecen.py
 ush/exglobal_prep_ocean_obs.py
```
Comment (aerorahul) [inline .gitignore]: These should be in `scripts/`

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN (hunk 5 above)
Comment (AndrewEichmann-NOAA) [reply jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN]: changed

File: sorc/link_workflow.sh (hunk 6 above)
Comment (AndrewEichmann-NOAA) [reply sorc/link_workflow.sh]: changed

File: .gitignore (hunk 7 above)
Comment (AndrewEichmann-NOAA) [reply .gitignore]: changed

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN [hunk 8]
```diff
@@ -5,6 +5,15 @@ source "${HOMEgfs}/ush/jjob_header.sh" -e "ocnanalecen" -c "base ocnanal ocnanal
 ##############################################
 # Set variables used in the script
 ##############################################
+# Ignore possible spelling error (nothing is misspelled)
+# shellcheck disable=SC2153
+GDATE=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} - ${assim_freq} hours")
+export gPDY=${GDATE:0:8}
+export gcyc=${GDATE:8:2}
+
+YMD=${gPDY} HH=${gcyc} generate_com -rx \
```
Comment (aerorahul) [inline jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN]: `generate_com` has been renamed to `declare_from_tmpl`

**Review by aerorahul — CHANGES_REQUESTED** [review]

File: jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN (hunk 8 above)
Comment (AndrewEichmann-NOAA) [reply jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN]: done

File: sorc/link_workflow.sh [hunk 9]
```diff
@@ -151,6 +151,8 @@ done
 
 cd "${HOMEgfs}/scripts" || exit 8
 ${LINK_OR_COPY} "${HOMEgfs}/sorc/ufs_utils.fd/scripts/exemcsfc_global_sfc_prep.sh" .
+${LINK_OR_COPY} "${HOMEgfs}/sorc/gdas.cd/scripts/exglobal_prep_ocean_obs.py"           .
+${LINK_OR_COPY} "${HOMEgfs}/sorc/gdas.cd/scripts/exgdas_global_marine_analysis_ecen.py"           .
```
Comment (WalterKolczynski-NOAA) [inline sorc/link_workflow.sh]: There should be a new (temporary) section for gdasapp script linking similar to what is done for all the executables and parm files:
```suggestion
if [[ -d "${HOMEgfs}/sorc/gdas.cd" ]]; then
  declare -a gdas_scripts=(exglobal_prep_ocean_obs.py \
                           exgdas_global_marine_analysis_ecen.py \
                           )
  for gdas_script in "${gdas_scripts[@]}" ; do
    ${LINK_OR_COPY} "${HOMEgfs}/sorc/gdas.cd/scripts/${gdas_script}" .
  done
fi

**Review by WalterKolczynski-NOAA — DISMISSED** [review]

File: sorc/link_workflow.sh (hunk 9 above)
Comment (AndrewEichmann-NOAA) [reply sorc/link_workflow.sh]: done

**Review by aerorahul — APPROVED** [review]
thanks for fixing the issues.

**Pull request merged**